lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1689,8 +1683,131 @@ void swift::performAbstractFuncDeclDiagnostics(TypeChecker &TC,
   AFD->getBody()->walk(VarDeclUsageChecker(TC, AFD));
 }
 
+/// Diagnose C style for loops.
+static void checkCStyleForLoop(TypeChecker &TC, const ForStmt *FS) {
+    // If we're missing semi-colons we'll already be erroring out, and this may not even have been intended as C-style.
+    if (FS->getSemicolon1Loc().isInvalid() || FS->getSemicolon2Loc().isInvalid())
+      return;
+    
+    InFlightDiagnostic diagnostic = TC.diagnose(FS->getStartLoc(), diag::deprecated_c_style_for_stmt);
+    
+    // Try to construct a fix it using for-each:
+    VarDecl *loopVar = nullptr;
+    Expr *startValue = nullptr;
+    Expr *endValue = nullptr;
+    bool strideByOne = false;
+    
+    // Verify that there is only one loop variable, and it is declared here.
+    auto initializers = FS->getInitializerVarDecls();
+    PatternBindingDecl *loopVarDecl = initializers.size() == 2 ? dyn_cast<PatternBindingDecl>(initializers[0]) : nullptr;
+    if (!loopVarDecl || loopVarDecl->getNumPatternEntries() != 1)
+      return;
+    startValue = loopVarDecl->getInit(0);
+    loopVar = dyn_cast<VarDecl>(initializers[1]);
+    
+    // Verify that the condition is a simple != or < comparison to the loop variable.
+    auto *Cond = FS->getCond().getPtrOrNull();
+    if (!Cond)
+      return;
+    auto callExpr = dyn_cast<CallExpr>(Cond);
+    if (!callExpr)
+      return;
+    auto dotSyntaxExpr = dyn_cast<DotSyntaxCallExpr>(callExpr->getFn());
+    if (!dotSyntaxExpr)
+      return;
+    auto binaryExpr = dyn_cast<BinaryExpr>(dotSyntaxExpr->getBase());
+    if (!binaryExpr)
+      return;
+    auto binaryFuncExpr = dyn_cast<DeclRefExpr>(binaryExpr->getFn());
+    if (!binaryFuncExpr)
+      return;
+    auto comparisonOpName = binaryFuncExpr->getDecl()->getNameStr();
+    if (comparisonOpName != "!=" && comparisonOpName != "<")
+      return;
+    auto args = binaryExpr->getArg()->getElements();
+    auto loadExpr = dyn_cast<LoadExpr>(args[0]);
+    if (!loadExpr)
+      return;
+    auto declRefExpr = dyn_cast<DeclRefExpr>(loadExpr->getSubExpr());
+    if (!declRefExpr)
+      return;
+    if (declRefExpr->getDecl() != loopVar)
```

Comment:
Please factor out the two major expression crawls here (determining whether we have a comparison/what is compares and determining whether we have an increment/what it increments). The whole function is getting very, very long and I fully expect that we'll be making those individual functions more complicated as we find more AST variations we need to match.